pkg/kds/mux/session.go

```diff
@@ -12,67 +12,93 @@ type Session interface {
 }
 
-func NewSession(peerID string, stream MultiplexStream, stop <-chan struct{}) Session {
+type MultiplexStream interface {
+	Send(*mesh_proto.Message) error
+	Recv() (*mesh_proto.Message, error)
+	Context() context.Context
+}
+
+func NewSession(peerID string, stream MultiplexStream) Session {
 	s := &session{
 		peerID: peerID,
-		done:   make(chan struct{}, 1),
-		err:    make(chan error, 1),
+		err:    make(chan error),
 		serverStream: &kdsServerStream{
-			requests:        make(chan *envoy_sd.DiscoveryRequest, 1),
-			MultiplexStream: stream,
+			ctx:          stream.Context(),
+			bufferStream: newBufferStream(),
 		},
 		clientStream: &kdsClientStream{
-			responses:       make(chan *envoy_sd.DiscoveryResponse, 1),
-			MultiplexStream: stream,
+			ctx:          stream.Context(),
+			bufferStream: newBufferStream(),
 		},
-		closed: int32(0),
 	}
 	go func() {
-		defer s.close()
-		if err := s.handle(stream, stop); err != nil {
-			s.err <- err
-		}
+		s.handleSend(stream)
+	}()
+	go func() {
+		s.handleRecv(stream)
 	}()
 	return s
 }
 
-func (s *session) handle(stream MultiplexStream, stop <-chan struct{}) error {
+func (s *session) handleRecv(stream MultiplexStream) {
 	for {
-		select {
-		case <-stop:
-			return nil
-		default:
-			if atomic.LoadInt32(&s.closed) == 1 {
-				return nil
-			}
-		}
-
 		msg, err := stream.Recv()
 		if err != nil {
-			return err
+			s.clientStream.bufferStream.close()
+			s.serverStream.bufferStream.close()
+			// Recv always finishes with either an EOF or another error
+			s.err <- err
+			return
 		}
 		switch v := msg.Value.(type) {
 		case *mesh_proto.Message_LegacyRequest:
-			s.serverStream.put(DiscoveryRequestV3(v.LegacyRequest))
+			msg = &mesh_proto.Message{Value: &mesh_proto.Message_Request{Request: DiscoveryRequestV3(v.LegacyRequest)}}
+			s.serverStream.bufferStream.put(msg)
 		case *mesh_proto.Message_Request:
-			s.serverStream.put(v.Request)
+			s.serverStream.bufferStream.put(msg)
 		case *mesh_proto.Message_LegacyResponse:
-			s.clientStream.put(DiscoveryResponseV3(v.LegacyResponse))
+			msg = &mesh_proto.Message{Value: &mesh_proto.Message_Response{Response: DiscoveryResponseV3(v.LegacyResponse)}}
+			s.clientStream.bufferStream.put(msg)
 		case *mesh_proto.Message_Response:
-			s.clientStream.put(v.Response)
+			s.clientStream.bufferStream.put(msg)
+		}
+	}
+}
+
+func (s *session) handleSend(stream MultiplexStream) {
+	kdsVersion := KDSVersion(stream.Context())
+	for {
+		select {
+		case itm, more := <-s.serverStream.bufferStream.sendBuffer:
+			if !more {
+				return
+			}
+			r := itm.msg
+			if kdsVersion == KDSVersionV2 && r != nil {
+				r = &mesh_proto.Message{Value: &mesh_proto.Message_LegacyResponse{LegacyResponse: DiscoveryResponseV2(r.GetResponse())}}
+			}
+			err := stream.Send(r)
+			itm.errChan <- err
+		case itm, more := <-s.clientStream.bufferStream.sendBuffer:
+			if !more {
+				return
+			}
+			r := itm.msg
+			if kdsVersion == KDSVersionV2 && r != nil {
```

Comment:
If `r` can actually be nil, then it still gets passed to `stream.Send`, which doesn't seem right?

Reply:
Actually forgot to remove the `r!=nil` which is not possible anymore